### ISSUE TITLE
PAML version installed via conda is an old one with a bug

### ISSUE THREAD

**Issue by jayoung** [bioconda/bioconda-recipes#38109]
hi there,

I'm a beginner at docker/conda/etc but I have used PAML a lot, so apologies if this is a mess-up on my end rather than in the conda recipe, but I think I'm onto something.

I had been installing PAML via conda within a docker container (`RUN conda install paml --channel bioconda`). (my container's base is `FROM continuumio/miniconda3:4.10.3`). PAML installs fine and the algorithm runs to completion within my container. 

However, one of the output files (rst) looks very odd, regardless of which input files I use.  I am pretty sure it's due to a bug that existed for a while in PAML, as described [here](https://groups.google.com/g/pamlsoftware/c/HXxqYBHYbRU/m/lLwe1V4CAwAJ).  The BEB probabilities also look odd in the mlc output file.

For the PAML nerds out there, there was a problem in the BEB algorithm in some versions of PAML in some versions that led to a bunch of 'nan' values in the rst file. The other values (non-nan) also seem off.

When I install from source, version 4.9j, obtained [here](http://abacus.gene.ucl.ac.uk/software/paml.html), the BEB results look good.

Likewise, version 4.10.6, obtained from [github](https://github.com/abacus-gene/paml) has 'good' BEB results.  On my system the precompiled executables on github for 4.10.6 don't work, but I can compile from source OK (I used gcc instead of cc).

If I understand right [this yaml](https://github.com/bioconda/bioconda-recipes/blob/master/recipes/paml/meta.yaml) tells us that conda is getting PAML source code from a galaxy depot.   It's a version 4.9 too. Unclear which sub-version it is, but I think it's older than 4.9j.

The solution on my end is simply not to use conda, but I did want to give a heads-up to anyone else using conda to install paml that you may get weird results.  

Is it possible to tell conda to use a newer PAML version rather than getting that one from Galaxy?  @ziheng-yang would be the perfect person to say which version is a stable release for this purpose.

thanks!

Janet

------------------------------------------------------------------- 

Dr. Janet Young 

Malik lab
http://research.fhcrc.org/malik/en.html
Division of Basic Sciences
Fred Hutchinson Cancer Center

-------------------------------------------------------------------

**Comment by ziheng-yang** [comment]
I am not familiar with conda.  if conda installs an earlier version of paml I imagine you can simply replace the executables from the more recent version.

Nan means not a number.  It means that the ml iteration has gone astray.  That can happen with any version, and you need to run the program again.

Best wishes, ziheng


From: jayoung ***@***.***>
Sent: Thursday, November 24, 2022 3:25 AM
To: bioconda/bioconda-recipes ***@***.***>
Cc: Yang, Ziheng ***@***.***>; Mention ***@***.***>
Subject: [bioconda/bioconda-recipes] PAML version installed via conda is an old one with a bug (Issue #38109)


hi there,

I'm a beginner at docker/conda/etc but I have used PAML a lot, so apologies if this is a mess-up on my end rather than in the conda recipe, but I think I'm onto something.

I had been installing PAML via conda within a docker container (RUN conda install paml --channel bioconda). (my container's base is FROM continuumio/miniconda3:4.10.3). PAML installs fine and the algorithm runs to completion within my container.

However, one of the output files (rst) looks very odd, regardless of which input files I use. I am pretty sure it's due to a bug that existed for a while in PAML, as described here<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgroups.google.com%2Fg%2Fpamlsoftware%2Fc%2FHXxqYBHYbRU%2Fm%2FlLwe1V4CAwAJ&data=05%7C01%7Cz.yang%40ucl.ac.uk%7C70eb9b593eda48f0b58908dacdcb8810%7C1faf88fea9984c5b93c9210a11d9a5c2%7C0%7C0%7C638048571291932701%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=wAoBuwuJLI650TTgg5hz76Lp2YxRlT4Fm3UPT2KPu98%3D&reserved=0>. The BEB probabilities also look odd in the mlc output file.

For the PAML nerds out there, there was a problem in the BEB algorithm in some versions of PAML in some versions that led to a bunch of 'nan' values in the rst file. The other values (non-nan) also seem off.

When I install from source, version 4.9j, obtained here<http://abacus.gene.ucl.ac.uk/software/paml.html>, the BEB results look good.

Likewise, version 4.10.6, obtained from github<https://github.com/abacus-gene/paml> has 'good' BEB results. On my system the precompiled executables on github for 4.10.6 don't work, but I can compile from source OK (I used gcc instead of cc).

If I understand right this yaml<https://github.com/bioconda/bioconda-recipes/blob/master/recipes/paml/meta.yaml> tells us that conda is getting PAML source code from a galaxy depot. It's a version 4.9 too. Unclear which sub-version it is, but I think it's older than 4.9j.

The solution on my end is simply not to use conda, but I did want to give a heads-up to anyone else using conda to install paml that you may get weird results.

Is it possible to tell conda to use a newer PAML version rather than getting that one from Galaxy? @ziheng-yang<https://github.com/ziheng-yang> would be the perfect person to say which version is a stable release for this purpose.

thanks!

Janet

________________________________

Dr. Janet Young

Malik lab
http://research.fhcrc.org/malik/en.html<https://eur01.safelinks.protection.outlook.com/?url=http%3A%2F%2Fresearch.fhcrc.org%2Fmalik%2Fen.html&data=05%7C01%7Cz.yang%40ucl.ac.uk%7C70eb9b593eda48f0b58908dacdcb8810%7C1faf88fea9984c5b93c9210a11d9a5c2%7C0%7C0%7C638048571291932701%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=VLSBD2lVE3mXERbq0WohzF8MnLXl%2BSDwPUs3w0AkE8M%3D&reserved=0>
Division of Basic Sciences
Fred Hutchinson Cancer Center

________________________________

**Comment by jayoung** [comment]
Thanks, Ziheng.   Do you have a preferred recent 'stable' version?   v4.10.6 maybe?  or v4.9j?   Thought I'd check in, in case the github-paml hosted versions are somehow more experimental than the versions you have on the paml website.

With the conda-paml install I was getting the 'nan' values in BEB every time - multiple runs, using 6 different alignments.  I haven't seen 'nan' with multiple runs of various other versions of paml, so I do think there's something systematically bad about the conda-paml version.   Conda-paml's output has this in the header 'paml version 4.9, March 2015'. It's possible it's some bad interaction with PAML and the rest of the docker container's environment.  In any case, other PAML versions are fine within the container, so the solution is just to update PAML.

all the best,

Janet

**Comment by jayoung** [comment]
aha - I figured out that the conda-paml I was using suffers from an old bug described [here](https://github.com/abacus-gene/paml/blob/master/doc/pamlHistory.txt)
Version 4.9g, December 2017
(*) codeml.  A bug caused the BEB calculation under the site model M8
(NSsites = 8) to be incorrect, with the program printing out warming
messages like "strange: f[ 5] = -0.0587063 very small."  This bug was
introduced in version 4.9b and affects versions 4.9b-f.  A different
bug was introduced in version 4.9f that causes the log likelihood
function under the site model M8 (NSsites = 8) to be calculated
incorrectly.  These are now fixed.

As well as the messages ("strange: f[ 5] = -0.0587063 very small.") many of the BEB results in the rst file are nan.

That bug has long-since been fixed.  The [meta.yaml to install paml](https://github.com/bioconda/bioconda-recipes/blob/deb762d67883578bc6f102c71c021d23bb6a4e33/recipes/paml/meta.yaml) gets the paml tarball from somewhere on depot.galaxyproject.org/.  

Newer versions without that bug are available from here https://github.com/abacus-gene/paml/releases
